### PR TITLE
b2: close upload connections that have not been used in a long time

### LINKED ISSUE #252: B2 backend does not clean up stale upload connections
B2 backend maintains a pool of upload URLs and associated connections, which do not get cleaned up after being established unless an error happens.

This means that if one uses high thread count (say, `threads=32`) with B2 backend, after a period of intensive I/O metadata upload may hang for hours while the backend tries all connections one by one, establishing that each of them does not work and closing them, waiting for 5 minutes between attempts in the back-off logic:

```
Jun 23 06:23:01 stratofortress mount.s4ql[2089141]: Dumping metadata...
Jun 23 06:23:01 stratofortress mount.s3ql[2089141]: ..objects..
Jun 23 06:23:01 stratofortress mount.s3ql[2089141]: ..blocks..
Jun 23 06:23:02 stratofortress mount.s3ql[2089141]: ..inodes..
Jun 23 06:23:02 stratofortress mount.s3ql[2089141]: ..inode_blocks..
Jun 23 06:23:03 stratofortress mount.s3ql[2089141]: ..symlink_targets..
Jun 23 06:23:03 stratofortress mount.s3ql[2089141]: ..names..
Jun 23 06:23:03 stratofortress mount.s3ql[2089141]: ..contents..
Jun 23 06:23:03 stratofortress mount.s3ql[2089141]: ..ext_attributes..
Jun 23 06:23:04 stratofortress mount.s3ql[2089141]: Compressing and uploading metadata...
Jun 23 06:23:10 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 3)...
Jun 23 06:23:10 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 4)...
Jun 23 06:23:10 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 5)...
Jun 23 06:23:11 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 6)...
Jun 23 06:23:12 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 7)...
Jun 23 06:23:14 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 8)...
Jun 23 06:23:17 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 9)...
Jun 23 06:23:22 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 10)...
Jun 23 06:23:35 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 11)...
Jun 23 06:24:06 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 12)...
Jun 23 06:25:05 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 13)...
Jun 23 06:26:33 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 14)...
Jun 23 06:29:41 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 15)...
Jun 23 06:35:06 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 16)...
Jun 23 06:40:19 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 17)...
Jun 23 06:45:52 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 18)...
Jun 23 06:50:59 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 19)...
Jun 23 06:57:27 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 20)...
Jun 23 07:04:49 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 21)...
Jun 23 07:11:37 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 22)...
Jun 23 07:16:48 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 23)...
Jun 23 07:23:41 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 24)...
Jun 23 07:29:06 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 25)...
Jun 23 07:36:26 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 26)...
Jun 23 07:42:11 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 27)...
Jun 23 07:47:40 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 28)...
Jun 23 07:53:26 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 29)...
Jun 23 08:00:31 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 30)...
Jun 23 08:07:38 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 31)...
Jun 23 08:13:44 stratofortress mount.s3ql[2089141]: Encountered ConnectionClosed (connection was interrupted), retrying ObjectW.close (attempt 32)...
Jun 23 08:19:17 stratofortress mount.s3ql[2089141]: Wrote 38.9 MiB of compressed metadata.
Jun 23 08:19:17 stratofortress mount.s3ql[2089141]: Cycling metadata backups...
Jun 23 08:19:17 stratofortress mount.s3ql[2089141]: Backing up old metadata...
```

S3QL should either verify the connections before using them, or schedule closure of each established connection after a period of inactivity.

### PR DESCRIPTION
**Pull request by intelfx**
B2 does not proactively clean up established upload connections unless
one gets re-used and an error happens.

If high I/O thread count is used, when a single-thread upload (such as
metadata upload) is performed after a period of intensive I/O, s3ql may
hang for hours trying all connections one-by-one with a back-off period
after each attempt.

Instead, track the last time a connection has been used and close it
before re-use if enough time had passed. This is not optimal; it would
be better to proactively close open connections after a set period of
time to avoid excessive use of file descriptors and other resources.

Fixes #252.

### CODE REVIEW

**Comment by intelfx** [comment]
@Nikratio help needed. This fixes the bug for me (and is sound, as far as I can tell), but in a suboptimal way.

It would be much better to close lingering connections proactively (e. g. set up some kind of an asynchronous one-shot timer every time a particular connection is used, and re-set the timer once it gets successfully reused), but I'm not sure how do I do that in S3QL given it's a mix of synchronous and asynchronous code.

File: src/s3ql/backends/b2/b2_backend.py [hunk 1]
```diff
@@ -662,6 +665,16 @@ def _invalidate_upload_url(self, upload_url_info):
 
         log.debug('upload urls left: %d', len(self.available_upload_url_infos))
 
+    # TODO: schedule to disconnect connections as we use them, don't leave lingering connections until next request comes
+    def _expire_upload_url_info(self, upload_url_info):
+        '''Disconnects an active connection that has been unused for too long.'''
+        if (upload_url_info['connection'] is not None
+                and upload_url_info['lastUsed'] != 0
+                and time.monotonic() > upload_url_info['lastUsed'] + 10):
```
Comment (Nikratio) [inline src/s3ql/backends/b2/b2_backend.py:673]: 10 seconds feels pretty random. Maybe use the tcp-timeout backend option instead? Or introduce a new option?

**Comment by Nikratio** [comment]
You should be able to use standard trio or standard threading facilities. Can you ask a more specific question of where you're stuck?

I would rather not set up a timer but have a background thread that periodically goes through all active connections and closes old ones.

Could you check if this affects the other backends as well? If so, then we should try to fix this for all backends instead of just the B2 backend. If it doesn't affect the other backends, then maybe we can adopt their solution to the B2 backend as well...

**Comment by Nikratio** [comment]
Closing this pull request for now - feel free to re-open if you continue work on this.

**Comment by intelfx** [comment]
Yeah, sorry, I got sidetracked with other stuff (and what I wrote works for me, etc, you know the rest ;)). I do plan to finish the work eventually.

**Pull request closed without merging**